versioned/storage/common/src/main/java/org/projectnessie/versioned/storage/common/logic/ReferenceLogicImpl.java

```diff
@@ -0,0 +1,640 @@
+
+      if (ref != null && ref.name().startsWith(INTERNAL_PREFIX)) {
+        // do not expose internal references
+        ref = null;
+      }
+      r.add(ref);
+    }
+    return r;
+  }
+
+  @Override
+  @Nonnull
+  @jakarta.annotation.Nonnull
+  public PagedResult<Reference, String> queryReferences(
+      @Nonnull @jakarta.annotation.Nonnull ReferencesQuery referencesQuery) {
+    Optional<PagingToken> pagingToken = referencesQuery.pagingToken();
+
+    StoreKey prefix = referencesQuery.referencePrefix().map(StoreKey::keyFromString).orElse(null);
+
+    StoreKey begin =
+        pagingToken
+            .map(PagingToken::token)
+            .map(ByteString::toStringUtf8)
+            .map(StoreKey::key)
+            .orElse(prefix);
+
+    StoreIndex<CommitOp> index = createRefsIndexSupplier().get();
+
+    return new QueryIter(index, prefix, begin, referencesQuery.prefetch());
+  }
+
+  private final class QueryIter extends AbstractIterator<Reference>
+      implements PagedResult<Reference, String> {
+    private final StoreIndex<CommitOp> index;
+    private final Iterator<StoreIndexElement<CommitOp>> base;
+    private final StoreKey prefix;
+
+    private QueryIter(
+        StoreIndex<CommitOp> index, StoreKey prefix, StoreKey begin, boolean prefetch) {
+      this.index = index;
+      this.prefix = prefix;
+      this.base = index.iterator(begin, null, prefetch);
+    }
+
+    @Override
+    protected Reference computeNext() {
+      while (true) {
+        if (!base.hasNext()) {
+          return endOfData();
+        }
+
+        StoreIndexElement<CommitOp> el = base.next();
+        StoreKey k = el.key();
+        if (prefix != null && !k.startsWith(prefix)) {
+          return endOfData();
+        }
+        String name = k.rawString();
+        Reference r = maybeRecover(name, persist.findReference(el.key().rawString()), () -> index);
+        if (r != null) {
+          return r;
+        }
+      }
+    }
+
+    @Nonnull
+    @jakarta.annotation.Nonnull
+    @Override
+    public PagingToken tokenForKey(String key) {
+      return key != null ? pagingToken(copyFromUtf8(key)) : emptyPagingToken();
+    }
+  }
+
+  @Override
+  @Nonnull
+  @jakarta.annotation.Nonnull
+  public Reference createReference(
+      @Nonnull @jakarta.annotation.Nonnull String name,
+      @Nonnull @jakarta.annotation.Nonnull ObjId pointer)
+      throws RefAlreadyExistsException, RetryTimeoutException {
+    checkArgument(!isInternalReferenceName(name));
+
+    while (true) {
+      CommitReferenceResult commitToIndex = commitCreateReference(name, pointer);
+      Reference reference = commitToIndex.reference;
+
+      switch (commitToIndex.kind) {
+        case ADDED_TO_INDEX:
+          checkState(!reference.deleted(), "internal error");
+          try {
+            return persist.addReference(reference);
+          } catch (RefAlreadyExistsException e) {
+            if (!e.reference().deleted()) {
```

Comment:
Nice catch!
So yes, `RefAlreadyExistsException` can return `null` and that's correct, because a concurrent reference-delete might have happened. So all call sites of `org.projectnessie.versioned.storage.common.exceptions.RefAlreadyExistsException` must use fetchReference()